src/cairoUtilFuncGen/inputArgCheck/inputCheck.ts

```diff
@@ -0,0 +1,267 @@
+import assert from 'assert';
+import {
+  AddressType,
+  ArrayType,
+  ASTNode,
+  BoolType,
+  DataLocation,
+  EnumDefinition,
+  FunctionCall,
+  FunctionStateMutability,
+  generalizeType,
+  getNodeType,
+  IntType,
+  SourceUnit,
+  StructDefinition,
+  TypeNode,
+  UserDefinedType,
+  VariableDeclaration,
+} from 'solc-typed-ast';
+import { FunctionStubKind } from '../../ast/cairoNodes';
+import { printTypeNode } from '../../utils/astPrinter';
+import {
+  CairoDynArray,
+  CairoFelt,
+  CairoType,
+  TypeConversionContext,
+} from '../../utils/cairoTypeSystem';
+import { NotSupportedYetError } from '../../utils/errors';
+import { createCairoFunctionStub, createCallToFunction } from '../../utils/functionGeneration';
+import { createIdentifier } from '../../utils/nodeTemplates';
+import { mapRange, narrowBigIntSafe, typeNameFromTypeNode } from '../../utils/utils';
+import { StringIndexedFuncGen } from '../base';
+import * as pathLib from 'path';
+
+export class InputCheckGen extends StringIndexedFuncGen {
+  sourceUnit: SourceUnit | undefined;
+  gen(node: VariableDeclaration, nodeInSourceUnit: ASTNode): FunctionCall {
+    const type = generalizeType(getNodeType(node, this.ast.compilerVersion))[0];
+    const functionInput = createIdentifier(node, this.ast);
+    this.sourceUnit = this.ast.getContainingRoot(nodeInSourceUnit);
+    const name = this.getOrCreate(type);
+    const functionStub = createCairoFunctionStub(
+      name,
+      [['ref_var', typeNameFromTypeNode(type, this.ast), DataLocation.CallData]],
```

Comment:
This should be a locationIfComplexType rather than just always being CallData, even for scalars